### DISCUSSION TITLE
Any way to get a variable from top level project into subproject?

### DISCUSSION THREAD

**Discussion by mahmnv** [mesonbuild/meson#12429] [General]
There is meson API for:                 
`subproject().get_variable(‘var_name’)`

is there something like:              
`superproject().get_variable(‘var_name’)`?

**Comment by xclaesse** [comment]
If you need to pass some setting from parent project that should be done with an option. Add an option to your subproject (meson_options.txt) and then pass a value from parent project with subproject('foo', default_options: ...).

**Reply by ayungman** [reply]
We need to pass variables not project settings, e.g `path` to something or any other variable that we build in the superproject meson

we look for a way to perform something like:
subproject().set_variable(‘var_name=var_value’)

**Reply by eli-schwartz** [reply]
I'm not entirely sure I understand the desired goal here.

In general, subprojects are conceived as a way to perform hermetically sealed, sandboxed dependency injection and vendoring. The preferred way to use it is via `dependency()` with overrides. The get_variable() approach is a legacy (although not deprecated) approach.

Getting a variable from the *parent* project and passing it to a vendored subproject feels like it's violating the ordering. Suddenly the subproject isn't really a subproject -- it knows it must be called only from a specific parent project and depends on the state (and internal implementation details) of the parent project's meson.build

Maybe it would help if I could see a real life example of where this would end up being used?

**Reply by mitchgrout** [reply]
I've had similar situations in which a third-party library requires some sort of non-trivial configuration file to compile, or perhaps requires a set of methods to be statically defined. A pattern you could try is within your subproject, require a dependency named `<subproject>-config`, and have the subproject build targets depend on that. With that, you can insert extra information (primarily extra sources and includes) as necessary. It's not particularly elegant, and rather limited; if you're in control of the subproject, it would be preferable to set it up so that you don't need to transmit structured data into it.
